src/github.com/matrix-org/dendrite/clientapi/auth/storage/devices/storage.go

```diff
@@ -15,23 +15,82 @@
 package devices
 
 import (
+	"database/sql"
+
 	"github.com/matrix-org/dendrite/clientapi/auth/authtypes"
+	"github.com/matrix-org/gomatrixserverlib"
 )
 
 // Database represents a device database.
 type Database struct {
-	// TODO
+	db      *sql.DB
+	devices devicesStatements
 }
 
 // NewDatabase creates a new device database
-func NewDatabase(dataSource string) (*Database, error) {
-	return &Database{}, nil
+func NewDatabase(dataSourceName string, serverName gomatrixserverlib.ServerName) (*Database, error) {
+	var db *sql.DB
+	var err error
+	if db, err = sql.Open("postgres", dataSourceName); err != nil {
+		return nil, err
+	}
+	d := devicesStatements{}
+	if err = d.prepare(db, serverName); err != nil {
+		return nil, err
+	}
+	return &Database{db, d}, nil
 }
 
 // GetDeviceByAccessToken returns the device matching the given access token.
+// Returns sql.ErrNoRows if no matching device was found.
 func (d *Database) GetDeviceByAccessToken(token string) (*authtypes.Device, error) {
-	// TODO: Actual implementation
+	// return d.devices.selectDeviceByToken(token) TODO: Figure out how to make integ tests pass
```

Comment:
^ Is this TODO still valid? I assume you got the integ tests passing since it passes CI?